tools/GulpCoverageUtils.ts

```diff
@@ -0,0 +1,129 @@
+        coverageOutputDir = path.resolve(coverageOutputDir);
+        modulesRoot = path.resolve(modulesRoot);
+
+        return GulpUtils.runNpmScript(modulesToTest, modulesRoot, GulpCoverageUtils.CoverageCommand)
+            .then(function(): Q.Promise<any> {
+                return Q.all<any>(modulesToTest.map(moduleName => {
+                    // this post coverage copy back is needed more for the scenario to support
+                    // coverage reports from a different machine (say MACOS)
+                    return GulpCoverageUtils.copyGeneratedCoverageJson(moduleName, modulesRoot, coverageOutputDir);
+                }));
+            })
+            .then(function(): Q.Promise<any> {
+                var coverageResults: string[] = [coverageOutputDir];
+                if (otherCoverageDir) {
+                    coverageResults.push(otherCoverageDir);
+                }
+                return GulpCoverageUtils.mergeAndGenerateReports(coverageResults, coverageOutputDir, modulesRoot);
+            });
+    }
+
+    private static mergeAndGenerateReports(coverageResultsPaths: string[], coverageOutputDir: string, modulesRoot: string): Q.Promise<any> {
+        return GulpCoverageUtils.collectCoverageJsons(coverageResultsPaths)
+            .then(function(coverageJsonsList: string[]): any {
+                return GulpCoverageUtils.normalizeCoveragePaths(coverageJsonsList, coverageOutputDir, modulesRoot);
+            })
+            .then(function(coverageJsonsList: string[]): void {
+                // collect all generated coverage.json and use them to remap to typescript sources
+                var loadCoverage: any = require("remap-istanbul/lib/loadCoverage");
+                var remap: any = require("remap-istanbul/lib/remap");
+                var writeReport: any = require("remap-istanbul/lib/writeReport");
+
+                var coverage = loadCoverage(coverageJsonsList);
+                var collector = remap(coverage, { basePath: "./" });
+                return writeReport(collector, GulpCoverageUtils.CoverageReportType, coverageOutputDir);
+            });
+    }
+
+    private static collectCoverageJsons(coverageResultsPaths: string[]): Q.Promise<string[]> {
+        // read all the directories and prepare a flat list of coverageXXX.json we need to remap
+        return Q.all<string[]>(coverageResultsPaths.map(coverageResultsPath => {
+            coverageResultsPath = path.resolve(coverageResultsPath);
+            return Q.denodeify(fs.readdir)(coverageResultsPath)
+                .then(function(files: string[]): string[] {
+                    return files.filter(file => {
+                        return file.toLowerCase().indexOf("coverage-") === 0 && path.extname(file) === ".json";
+                    }).map(file => path.join(coverageResultsPath, file));
+                });
+        })).spread(function(...args: string[][]): string[] {
```

Comment:
Why spread and then use ...args? You can just use a .then taking the `string[][]` as its first argument